embassy-nxp/src/dma/lpc55.rs

```diff
@@ -0,0 +1,371 @@
+        // when the current descriptor is exhausted.
+        // Reloading allows ping-pong and linked transfers.
+        w.set_reload(false);
+        // There is no hardware distinction between interrupt A and B.
+        // They can be used by software to assist with more complex descriptor usage.
+        // By convention, interrupt A may be used when only one interrupt flag is needed.
+        w.set_setinta(true);
+        w.set_setintb(false);
+        w.set_width(data_size);
+        w.set_srcinc(if incr_src {
+            dma::vals::Srcinc::WIDTH_X_1
+        } else {
+            dma::vals::Srcinc::NO_INCREMENT
+        });
+        w.set_dstinc(if incr_dest {
+            dma::vals::Dstinc::WIDTH_X_1
+        } else {
+            dma::vals::Dstinc::NO_INCREMENT
+        });
+        // Total number of transfers to be performed, minus 1 encoded.
+        w.set_xfercount((len as u16) - 1);
+        // Before triggering the channel, it has to be enabled.
+        w.set_swtrig(false);
+    });
+
+    compiler_fence(Ordering::SeqCst);
+    DMA0.enableset0().write(|w| w.set_ena(1 << ch.number()));
+    DMA0.intenset0().write(|w| w.set_inten(1 << ch.number()));
+
+    compiler_fence(Ordering::SeqCst);
+    // Start transfer.
+    DMA0.settrig0().write(|w| w.set_trig(1 << ch.number()));
+    compiler_fence(Ordering::SeqCst);
+    Transfer::new(ch)
+}
+
+/// DMA transfer driver.
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct Transfer<'a, C: Channel> {
+    channel: Peri<'a, C>,
+}
+
+impl<'a, C: Channel> Transfer<'a, C> {
+    pub(crate) fn new(channel: Peri<'a, C>) -> Self {
+        Self { channel }
+    }
+}
+
+impl<'a, C: Channel> Drop for Transfer<'a, C> {
+    fn drop(&mut self) {
+        DMA0.enableclr0().write(|w| w.set_clr(1 << self.channel.number()));
+        while (DMA0.busy0().read().bsy() & (1 << self.channel.number())) != 0 {}
+        DMA0.abort0().write(|w| w.set_abortctrl(1 << self.channel.number()));
+    }
+}
+
+impl<'a, C: Channel> Unpin for Transfer<'a, C> {}
+impl<'a, C: Channel> Future for Transfer<'a, C> {
+    type Output = ();
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // We need to register/re-register the waker for each poll because any
+        // calls to wake will deregister the waker.
+        CHANNEL_WAKERS[self.channel.number() as usize].register(cx.waker());
+        // Check if it is busy or not.
+        if (DMA0.busy0().read().bsy() & (1 << self.channel.number())) != 0 {
+            Poll::Pending
+        } else {
+            Poll::Ready(())
+        }
+    }
+}
+
+// Total number of channles including both DMA0 and DMA1.
+// In spite of using only DMA0 channels, the descriptor table
+// should be of this size.
+pub(crate) const CHANNEL_COUNT: usize = 32;
+
+static CHANNEL_WAKERS: [AtomicWaker; CHANNEL_COUNT] = [const { AtomicWaker::new() }; CHANNEL_COUNT];
+
+// See section 22.5.2 (table 450)
```

Comment:
Should probably specify `UM11126, Rev. 2.8` so we know which specific document is saying this.